build/BuildContext.cs

```diff
@@ -13,35 +16,61 @@ public enum ProjectType
 
 public class BuildContext : FrostingContext
 {
+    public static string VersionBase = "1.0.0";
+    public static readonly Regex VersionRegex = new(@"^v\d+.\d+.\d+", RegexOptions.Compiled | RegexOptions.IgnoreCase);
     public static readonly string DefaultRepositoryUrl = "https://github.com/MonoGame/MonoGame";
-    public static readonly string DefaultBaseVersion = "3.8.2";
 
     public BuildContext(ICakeContext context) : base(context)
     {
         var repositoryUrl = context.Argument("build-repository", DefaultRepositoryUrl);
         var buildConfiguration = context.Argument("build-configuration", "Release");
         BuildOutput = context.Argument("build-output", "artifacts");
-        Version = context.Argument("build-version", DefaultBaseVersion + ".1-develop");
         NuGetsDirectory = $"{BuildOutput}/NuGet/";
 
+        var tags = GitAliases.GitTags(context, ".");
+        foreach (var tag in tags)
+        {
+            if (VersionRegex.IsMatch(tag.FriendlyName))
+            {
+                VersionBase = tag.FriendlyName[1..];
+            }
+        }
+
         if (context.BuildSystem().IsRunningOnGitHubActions)
         {
             var workflow = context.BuildSystem().GitHubActions.Environment.Workflow;
             repositoryUrl = $"https://github.com/{workflow.Repository}";
 
+            if (workflow.RefType == GitHubActionsRefType.Tag)
+            {
+                var baseVersion = workflow.RefName;
+                if (Regex.IsMatch(baseVersion, @"v\d+.\d+.\d+"))
+                {
+                    VersionBase = baseVersion[1..];
+                }
+                else
+                {
+                    throw new Exception($"Invalid tag: {baseVersion}");
+                }
+            }
+
             if (workflow.Repository != "MonoGame/MonoGame")
             {
-                Version = $"{DefaultBaseVersion}.{workflow.RunNumber}-{workflow.RepositoryOwner}";
+                Version = $"{VersionBase}.{workflow.RunNumber}-{workflow.RepositoryOwner}";
             }
             else if (workflow.RefType == GitHubActionsRefType.Branch && workflow.RefName != "refs/heads/master")
             {
-                Version = $"{DefaultBaseVersion}.{workflow.RunNumber}-develop";
+                Version = $"{VersionBase}.{workflow.RunNumber}-develop";
```

Comment:
the `develop` shouldn't be hard-coded and needs to the parameterised, so we can swap that to `preview` when build from non `master` branch for the preview release.

Reply:
I've modified the code so that if non master branch is tagged, it will pick up version from the tag fully.

Reply:
OK LGTM